### ISSUE TITLE
Can you please help me to understand the couple places in train.py that are coco specific

### ISSUE THREAD

**Issue by davesargrad** [ultralytics/yolov5#3640]
## ❔Question
In the following we see a portion of code that is only run (at the end of a training session), in the event that the input dataset was COCO.

![image](https://user-images.githubusercontent.com/48793689/122212958-6cb20180-ce76-11eb-9301-f791207cfe0c.png)

Can you please help me to understand the reason for this code? I can see using train.py to train a model from scratch, using a dataset other than coco. Also given that I've studied the code, there really isn't much in the code that is COCO specific.

What is the specific need for such COCO-specific portions of the code?

Thanks in advance.

## Additional context
We also see coco in one or two other places:

![image](https://user-images.githubusercontent.com/48793689/122213357-dd591e00-ce76-11eb-80bb-ac5018ab16f9.png)

![image](https://user-images.githubusercontent.com/48793689/122213540-11ccda00-ce77-11eb-95d8-770dd44225d2.png)

![image](https://user-images.githubusercontent.com/48793689/122214132-ba7b3980-ce77-11eb-9c34-edaffd1abaa6.png)

**Comment by glenn-jocher** [comment]
@davesargrad this code simply gathers stats used for R&D automatically after each training.

<img width="1447" alt="Screenshot 2021-06-18 at 11 27 12" src="https://user-images.githubusercontent.com/26833433/122539689-27cdcc80-d028-11eb-9492-16834bb8e65d.png">

**Comment by davesargrad** [comment]
@glenn-jocher 
Hi. Thanks for your answer
Clearly I didn't ask my question properly. I fully understand that this is code to gather statistics after training.

My real question is why does that statistics collection mechanism have to be coco specific?

train.py, as it stands, is so close to being fully generalized and disconnected from the underlying datasets that are used to train a model.

I would have thought that the statistics of interest can still be collected, without making the code tie specifically to "is this COCO data"? 

In other words, the same statistics will be of value when training with datasets other than COCO, and with a seemingly simple generalization of the code they can be collected.

**Comment by glenn-jocher** [comment]
@davesargrad yeah that's a good idea! 

The main purpose of duplicating the testing code here is that this particular code treats the model as independent from training as it loads it directly from the checkpoint and passes the settings used to reproduce mAP in the readme, so theoretically this code should give you the same results as if you later ran `python test.py --weights path/to/best.pt` at a later point in time, but yes we could just as well generalize it to be dataset independent.
